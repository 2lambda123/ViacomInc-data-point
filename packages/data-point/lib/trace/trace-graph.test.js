/* eslint-env jest */

const TraceGraph = require('./trace-graph')
const set = require('lodash/set')
const { format } = require('util')

function createNode (index) {
  const time = index * 1e9
  return {
    id: `id${index}`,
    time,
    timeStart: time,
    duration: time + 20000,
    parent: null,
    reducer: {
      id: `reducerId${index}`,
      name: `reducerName${index}`
    }
  }
}

function createTraceGraph () {
  const traceGraph = Array(4)
    .fill(null)
    .map((val, index) => {
      return createNode(index)
    })

  traceGraph[1].parent = traceGraph[0]
  traceGraph[2].parent = traceGraph[0]
  traceGraph[3].parent = traceGraph[2]

  return traceGraph
}

function createGraph () {
  const traceGraph = createTraceGraph()
  const root = traceGraph[0]
  const graphAcc = { timeStart: root.timeStart, maxNestingLevel: 0 }
  TraceGraph.createTree(root, traceGraph, 0, graphAcc)
  return {
    root,
    graphAcc
  }
}

describe('createTraceNodeLabel', () => {
  it('should use name when available', () => {
    const node = {}
    set(node, 'id', 'nodeId')
    set(node, 'reducer.id', 'reducerId')
    set(node, 'reducer.name', 'reducerName')
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual(
      'reducerId:nodeId (reducerName)'
    )
  })
  it('should not set name when not available', () => {
    const node = {}
    set(node, 'id', 'nodeId')
    set(node, 'reducer.id', 'reducerId')
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual('reducerId:nodeId')
  })
  it('should use reducer type when id is not available', () => {
    const node = {}
    set(node, 'id', 'nodeId')
    set(node, 'reducer.type', 'reducerType')
    expect(TraceGraph.createTraceNodeLabel(node)).toEqual('reducerType:nodeId')
  })
})

describe('createTree', () => {
  it('should create parent child relationship', () => {
    const result = createGraph()
    expect(result.root).toMatchSnapshot()
  })

  it('should set maxNestingLevel', () => {
    const result = createGraph()
    expect(result.graphAcc.maxNestingLevel).toEqual(3)
  })
})

describe('logGraph', () => {
  let mockConsoleLog
  afterEach(() => {
    mockConsoleLog.mockRestore()
  })

  it('should log the tree', () => {
    const logs = []
    mockConsoleLog = jest
      .spyOn(console, 'log')
      .mockImplementation((...args) => {
        logs.push(format(...args))
      })

    TraceGraph.logGraph(createGraph().root)
    const logOutput = logs.join('\n')
    expect(logOutput).toMatchSnapshot()
  })
})

describe('writeTraceGraph', () => {
  let mockDateNow
  let mockWriteFileP
  beforeEach(() => {
    mockDateNow = jest.spyOn(Date, 'now').mockImplementation(() => {
      return 123
    })
    mockWriteFileP = jest
      .spyOn(TraceGraph, 'writeFileP')
      .mockImplementation(() => {
        return Promise.resolve(true)
      })
  })
  afterEach(() => {
    mockDateNow.mockRestore()
    mockWriteFileP.mockRestore()
  })
  it('should write graph to disk', () => {
    const traceGraph = createTraceGraph()
    return TraceGraph.writeTraceGraph(traceGraph).then(result => {
      expect(mockWriteFileP.mock.calls[0]).toMatchSnapshot()
    })
  })
})
