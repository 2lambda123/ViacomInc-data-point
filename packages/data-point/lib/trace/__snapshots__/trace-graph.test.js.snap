// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTree should create parent child relationship 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "duration": 1000020000,
      "durationMs": 1.00002,
      "id": "id1",
      "label": "reducerId1:id1 (reducerName1)",
      "nestingLevel": 1,
      "parent": [Circular],
      "reducer": Object {
        "id": "reducerId1",
        "name": "reducerName1",
      },
      "time": 1000000000,
      "timeStart": 1000000000,
      "timelineStart": 1000000000,
      "timelineStartMs": 1,
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "duration": 3000020000,
          "durationMs": 3.00002,
          "id": "id3",
          "label": "reducerId3:id3 (reducerName3)",
          "nestingLevel": 2,
          "parent": [Circular],
          "reducer": Object {
            "id": "reducerId3",
            "name": "reducerName3",
          },
          "time": 3000000000,
          "timeStart": 3000000000,
          "timelineStart": 3000000000,
          "timelineStartMs": 3,
        },
      ],
      "duration": 2000020000,
      "durationMs": 2.00002,
      "id": "id2",
      "label": "reducerId2:id2 (reducerName2)",
      "nestingLevel": 1,
      "parent": [Circular],
      "reducer": Object {
        "id": "reducerId2",
        "name": "reducerName2",
      },
      "time": 2000000000,
      "timeStart": 2000000000,
      "timelineStart": 2000000000,
      "timelineStartMs": 2,
    },
  ],
  "duration": 20000,
  "durationMs": 0.00002,
  "id": "id0",
  "label": "reducerId0:id0 (reducerName0)",
  "nestingLevel": 0,
  "parent": null,
  "reducer": Object {
    "id": "reducerId0",
    "name": "reducerName0",
  },
  "time": 0,
  "timeStart": 0,
  "timelineStart": 0,
  "timelineStartMs": 0,
}
`;

exports[`logGraph should log the tree 1`] = `
"0 reducerId0:id0 (reducerName0) S: 0.000ms T: 0.000ms
  1 reducerId1:id1 (reducerName1) S: 1.000ms T: 1.000ms
  1 reducerId2:id2 (reducerName2) S: 2.000ms T: 2.000ms
    2 reducerId3:id3 (reducerName3) S: 3.000ms T: 3.000ms"
`;

exports[`writeTraceGraph should write graph to disk 1`] = `
Array [
  "data-point-trace-123.json",
  "{
  \\"id\\": \\"id0\\",
  \\"time\\": 0,
  \\"timeStart\\": 0,
  \\"duration\\": 20000,
  \\"parent\\": null,
  \\"reducer\\": {
    \\"id\\": \\"reducerId0\\",
    \\"name\\": \\"reducerName0\\"
  },
  \\"durationMs\\": 0.00002,
  \\"timelineStart\\": 0,
  \\"timelineStartMs\\": 0,
  \\"nestingLevel\\": 0,
  \\"label\\": \\"reducerId0:id0 (reducerName0)\\",
  \\"children\\": [
    {
      \\"id\\": \\"id1\\",
      \\"time\\": 1000000000,
      \\"timeStart\\": 1000000000,
      \\"duration\\": 1000020000,
      \\"parent\\": \\"[Circular ~]\\",
      \\"reducer\\": {
        \\"id\\": \\"reducerId1\\",
        \\"name\\": \\"reducerName1\\"
      },
      \\"durationMs\\": 1.00002,
      \\"timelineStart\\": 1000000000,
      \\"timelineStartMs\\": 1,
      \\"nestingLevel\\": 1,
      \\"label\\": \\"reducerId1:id1 (reducerName1)\\",
      \\"children\\": []
    },
    {
      \\"id\\": \\"id2\\",
      \\"time\\": 2000000000,
      \\"timeStart\\": 2000000000,
      \\"duration\\": 2000020000,
      \\"parent\\": \\"[Circular ~]\\",
      \\"reducer\\": {
        \\"id\\": \\"reducerId2\\",
        \\"name\\": \\"reducerName2\\"
      },
      \\"durationMs\\": 2.00002,
      \\"timelineStart\\": 2000000000,
      \\"timelineStartMs\\": 2,
      \\"nestingLevel\\": 1,
      \\"label\\": \\"reducerId2:id2 (reducerName2)\\",
      \\"children\\": [
        {
          \\"id\\": \\"id3\\",
          \\"time\\": 3000000000,
          \\"timeStart\\": 3000000000,
          \\"duration\\": 3000020000,
          \\"parent\\": \\"[Circular ~.children.1]\\",
          \\"reducer\\": {
            \\"id\\": \\"reducerId3\\",
            \\"name\\": \\"reducerName3\\"
          },
          \\"durationMs\\": 3.00002,
          \\"timelineStart\\": 3000000000,
          \\"timelineStartMs\\": 3,
          \\"nestingLevel\\": 2,
          \\"label\\": \\"reducerId3:id3 (reducerName3)\\",
          \\"children\\": []
        }
      ]
    }
  ],
  \\"maxNestingLevel\\": 3
}",
  "utf8",
]
`;
